dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/programrule/ProgramRuleAssignActionTest.java

```diff
@@ -0,0 +1,181 @@
+
+class ProgramRuleAssignActionTest extends TrackerTest
+{
+    private static final String EVENT_UID = "D9PbzJY8bJO";
+
+    @Autowired
+    private TrackerImportService trackerImportService;
+
+    @Autowired
+    private ProgramRuleService programRuleService;
+
+    @Autowired
+    private ProgramRuleActionService programRuleActionService;
+
+    @Autowired
+    private ProgramRuleVariableService programRuleVariableService;
+
+    @Autowired
+    private SystemSettingManager systemSettingManager;
+
+    private Program program;
+
+    private DataElement dataElement1;
+
+    @Override
+    public void initTest()
+        throws IOException
+    {
+        ObjectBundle bundle = setUpMetadata( "tracker/simple_metadata.json" );
+        program = bundle.getPreheat().get( PreheatIdentifier.UID, Program.class, "BFcipDERJnf" );
+        dataElement1 = bundle.getPreheat().get( PreheatIdentifier.UID, DataElement.class, "DATAEL00001" );
+        DataElement dataElement2 = bundle.getPreheat().get( PreheatIdentifier.UID, DataElement.class, "DATAEL00002" );
+        ProgramRuleVariable programRuleVariable = createProgramRuleVariableWithDataElement( 'A', program,
+            dataElement2 );
+        programRuleVariableService.addProgramRuleVariable( programRuleVariable );
+
+        injectAdminUser();
+
+        assignProgramRule();
+        TrackerImportParams params = fromJson( "tracker/programrule/tei_enrollment_completed_event.json" );
+
+        trackerImportService.importTracker( params );
+    }
+
+    @Test
+    void shouldImportWithWarningWhenDataElementWithSameValueIsAssignedByAssignRule()
+        throws IOException
+    {
+        TrackerImportParams params = fromJson( "tracker/programrule/event_update_datavalue_same_value.json" );
+        params.setImportStrategy( TrackerImportStrategy.CREATE_AND_UPDATE );
+
+        ImportReport importReport = trackerImportService.importTracker( params );
+
+        Assertions.assertHasOnlyWarningsAndNoErrors( importReport.getValidationReport(),
+            new Warning(
+                "Generated by program rule (`ProgramRulF`) - DataElement `DATAEL00001` is being replaced in event `D9PbzJY8bJO`",
+                E1308.name(), EVENT.name(), EVENT_UID ) );
+    }
+
+    @Test
+    void shouldNotImportWhenDataElementWithDifferentValueIsAssignedByAssignRule()
+        throws IOException
+    {
+        TrackerImportParams params = fromJson( "tracker/programrule/event_update_datavalue_different_value.json" );
+        params.setImportStrategy( TrackerImportStrategy.CREATE_AND_UPDATE );
+
+        ImportReport importReport = trackerImportService.importTracker( params );
+
+        assertHasOnlyErrors( importReport, E1307 );
+    }
+
+    @Test
+    void shouldImportWithWarningWhenDataElementWithDifferentValueIsAssignedByAssignRuleAndOverwriteKeyIsTrue()
+        throws IOException
+    {
+        systemSettingManager.saveSystemSetting( SettingKey.RULE_ENGINE_ASSIGN_OVERWRITE, true );
```

Comment:
Do we need to "cleanup" this system setting change? So we don't run into a flaky test in the future.

Reply:
Not really, this test class is extending TrackerTest that is extending SingleSetupIntegrationTestBase that is running tests inside a transaction. Enything changed inside a test is rollbacked when the test is done